celer/group_fast.pyx

```diff
@@ -25,29 +25,34 @@ cdef:
 @cython.cdivision(True)
 cpdef floating primal_grplasso(
         floating alpha, floating[:] R, int[::1] grp_ptr,
-        int[::1] grp_indices, floating[:] w):
+        int[::1] grp_indices, floating[:] w, floating[:] weights):
     cdef floating nrm = 0.
     cdef int j, k, g
     cdef int n_samples = R.shape[0]
     cdef int n_groups = grp_ptr.shape[0] - 1
     cdef floating p_obj = fnrm2(&n_samples, &R[0], &inc) ** 2 / (2 * n_samples)
+
     for g in range(n_groups):
         nrm = 0.
+
         for k in range(grp_ptr[g], grp_ptr[g + 1]):
             j = grp_indices[k]
             nrm += w[j] ** 2
-        p_obj += alpha * sqrt(nrm)
+        p_obj += sqrt(nrm) * weights[g]
+
+    p_obj *= alpha
```

Comment:
This induces an error, when doing this you also multiply the datafit by alpha

Reply:
Fixed!
Yet, I am still a bit confused about that since ``a*(b+c+d) = a*b+a*c+a*d`` (basic math).
Also, I believe that the left-hand side way of computing (3 operations) is more effective than the right-hand side (5 operations). More precisely if we have a sum with ``n`` terms, the left-hand side needs ``n`` operation to be computed whereas the right-hand side needs ``n + n - 1``.

Reply:
1) this is not a bottleneck of the computation O(p) vs O(np) for the critical part.
2) as I said, with your method the datafit is multiplied by alpha. It should not.